data/tutorials/lg_08_error_handling.md

```diff
@@ -125,67 +191,549 @@ let () =
+
+```shell
+> echo "(Obj.magic () : int array).(0)" > foo.ml
+> ocamlopt foo.ml
+> ./a.out
+Segmentation fault (core dumped)
+```
+
+### Language Bugs
+
+When a crash isn't coming from:
+* A limitation of the native code compiler
+* `Obj.magic`
+
+It may be a language bug. It happens. Here is what to do when this is suspected:
+
+1. Make sure the crash affects both compilers: bytecode and native
+1. Write a self-contained and minimal proof-of-concept code which does nothing
+   but triggering the crash
+1. File an issue in the [OCaml Bug Tracker in
+   GitHub](https://github.com/ocaml/ocaml/issues)
+
+Here is an example of such a bug: https://github.com/ocaml/ocaml/issues/7241
+
+### Safe vs. Unsafe Functions
+
+Uncaught exceptions raise runtime crashes. Therefore, there is a tendency to use
+the following terminology:
+* Function raising exceptions: Unsafe
+* Function handling errors in data: Safe
+
+The main means to write such kind of safe error handling functions is to use
+either `Option` (next section) or `Result` (following section).
+
+## Using the `Option` Type for Errors
+
+The `Option` module provides the first alternative to exceptions. The `'a
+option` datatype allows to express either the availability of data for instance
+`Some 42` or the absence of data using `None`, which can represent an error.
+
+Using `Option` it is possible to write functions that return `None` instead of
+throwing an exception.
+```ocaml
+let div_opt m n =
+  try Some (m / n) with
+    Division_by_zero -> None
+
+let find_opt p l =
+  try Some (List.find p l) with
+    Not_found -> None
+```
+We can try those functions:
+
+```ocaml
+# 1 / 0;;
+Exception: Division_by_zero.
+# div_opt 42 2;;
+- : int option = Some 24
+# div_opt 42 0;;
+- : int option = None
+# List.find (fun x -> x mod 2 = 0) [1; 3; 5];;
+Exception: Not_found.
+# find_opt (fun x -> x mod 2 = 0) [1; 3; 4; 5];;
+- : int option = Some 4
+# find_opt (fun x -> x mod 2 = 0) [1; 3; 5];;
+- : int option = None
+```
+
+This can even be turned into a higher-order generic function:
+```ocaml
+# let try_opt f x = try Some (f x) with _ -> None;;
+val try_opt : ('a -> 'b) -> 'a -> 'b option = <fun>
+```
 
-## Result Type
+It tends to be considered good practice nowadays when a function can fail in
+cases that are not bugs (i.e., not `assert false`, but network failures, keys
+not present, etc.) to return a more explicit type such as `'a option` or `('a,
+'b) result` (see next section).
 
-The Stdlib module contains the following type:
+### Naming Conventions
+
+There are two naming conventions to have two versions of the same partial
+function: one raising exception, the other returning an option. In the above
+examples, the convention of the standard library is to add an `_opt` suffix to
+name of the version of the function that returns an option instead of raising
+exceptions.
+```ocaml
+val find: ('a -> bool) -> 'a list -> 'a
+(** @raise Not_found *)
+val find_opt: ('a -> bool) -> 'a list -> 'a option
```

Comment:
Is this code meant to be examples of what they can write, or should it be something they can test right now?

If I add the `;;` to the end, I get this response:
```
Error: Value declarations are only allowed in signatures
```